### PR TITLE
Code Example: Denoising diffusion implicit models

### PR DESCRIPTION
**Pull request by beresandras**
Added a detailed example on denoising diffusion generative models, using the Oxford Flowers dataset.

Results gif [here](https://i.imgur.com/gyc8eZX.gif)

### CODE REVIEW

File: examples/generative/ddim.py [hunk 1]
```diff
@@ -0,0 +1,839 @@
...
+
+        # KID is estimated per batch and is averaged across batches
+        self.kid_tracker = keras.metrics.Mean()
+
+        # a pretrained InceptionV3 is used without its classification layer
+        # transform the pixel values to the 0-255 range, then use the same
+        # preprocessing as during pretraining
+        self.encoder = keras.Sequential(
+            [
+                layers.InputLayer(input_shape=(image_size, image_size, 3)),
```
Comment (fchollet) [inline examples/generative/ddim.py:194]: Prefer using `keras.Input(shape=(image_size, image_size, 3))`

File: examples/generative/ddim.py [hunk 2]
```diff
@@ -0,0 +1,839 @@
...
+                layers.Rescaling(255.0),
+                layers.Resizing(height=kid_image_size, width=kid_image_size),
+                layers.Lambda(keras.applications.inception_v3.preprocess_input),
+                keras.applications.InceptionV3(
```
Comment (fchollet) [inline examples/generative/ddim.py:198]: What motivates the choice of InceptionV3 (which is quite heavy)?

File: examples/generative/ddim.py [hunk 3]
```diff
@@ -0,0 +1,839 @@
...
+* We disable the learnable center and scale parameters of the batch normalization layers,
+since the following convolution layers make them redundant.
+* We initialize the last convolution's kernel to all zeros as a good practice, making the
+network predict only zeros after initialization, which is the mean of its targets. This
+will improve behaviour at the start of training and make the mean squared error loss
+start at exactly 1.
+"""
+
+
+def get_network():
```
Comment (fchollet) [inline examples/generative/ddim.py:297]: Following the closure idea, it would be more consistent to give this is a capitalized name

File: examples/generative/ddim.py [hunk 4]
```diff
@@ -0,0 +1,839 @@
...
+            x, skips = x
+            x = layers.UpSampling2D(size=2, interpolation="bilinear")(x)
+            for _ in range(block_depth):
+                x = layers.Concatenate()([x, skips.pop()])
+                x = ResidualBlock(width)(x)
+            return x
+
+        return apply
+
+    noisy_images = keras.Input(shape=(image_size, image_size, 3))
```
Comment (fchollet) [inline examples/generative/ddim.py:353]: image_size seems like it should be a constructor argument.

File: examples/generative/ddim.py [hunk 5]
```diff
@@ -0,0 +1,839 @@
...
+
+
+"""
+## Training
+"""
+
+# create and compile the model
+model = DiffusionModel()
+model.compile(
+    optimizer=tfa.optimizers.AdamW(
```
Comment (fchollet) [inline examples/generative/ddim.py:644]: Prefer using `keras.optimizers.experimental.AdamW` (which will soon be `keras.optimizers.AdamW`, in 2.10)

File: examples/generative/ddim.py [hunk 6]
```diff
@@ -0,0 +1,839 @@
...
+        plt.close()
+
+
+"""
+## Training
+"""
+
+# create and compile the model
+model = DiffusionModel()
+model.compile(
```
Comment (fchollet) [inline examples/generative/ddim.py:643]: Random question -- do you get better performance if you add `jit_compile=True`?

**Review by fchollet — COMMENTED** [review]
Thank you for the contribution! We definitely needed a diffusion model example. It is beautifully written and very readable 👍

File: examples/generative/ddim.py [hunk 7]
```diff
@@ -0,0 +1,839 @@
...
+
+"""
+
+
+class KID(keras.metrics.Metric):
+    def __init__(self, name="kid", **kwargs):
+        super().__init__(name=name, **kwargs)
+
+        # KID is estimated per batch and is averaged across batches
+        self.kid_tracker = keras.metrics.Mean()
```
[the remark this answers is not in the document]
Comment (beresandras) [reply examples/generative/ddim.py:190]: The KID class subclasses keras.metrics.Metric, and is named ("kid"). This mean tracker metric is never logged/shown explicitly. Should I still name it?

File: examples/generative/ddim.py (hunk 6 above)
Comment (beresandras) [reply examples/generative/ddim.py:647]: I get an error:
error message

And a similar one for bilinear upsampling.

File: examples/generative/ddim.py (hunk 2 above)
Comment (beresandras) [reply examples/generative/ddim.py:201]: A smaller network, like a pretrained MobileNet, could probably work as well. But InceptionV3 is what is commonly used in FID and KID metrics, they even have it in their names. I reduce its resolution from 299 to 75 to be lighter. I did the same in a GAN code example: https://github.com/keras-team/keras-io/blob/master/examples/generative/gan_ada.py#L221

File: examples/generative/ddim.py (hunk 3 above)
Comment (beresandras) [reply examples/generative/ddim.py:300]: Fair, but this returns a model object, not a function like the closures.

File: examples/generative/ddim.py [hunk 8]
```diff
@@ -0,0 +1,839 @@
...
+            )
+        )
+        angular_speeds = 2.0 * math.pi * frequencies
+        embeddings = tf.concat(
+            [tf.sin(angular_speeds * x), tf.cos(angular_speeds * x)],
+            axis=3,
+        )
+        return embeddings
+
+    def ResidualBlock(width):
```
[the remark this answers is not in the document]
Comment (beresandras) [reply examples/generative/ddim.py:317]: Should I put the sinusoidal embedding outside as well?

File: examples/generative/ddim.py [hunk 9]
```diff
@@ -0,0 +1,839 @@
...
+    noise_variances = keras.Input(shape=(1, 1, 1))
+
+    e = layers.Lambda(sinusoidal_embedding)(noise_variances)
+    e = layers.UpSampling2D(size=image_size, interpolation="nearest")(e)
+
+    x = layers.Conv2D(widths[0], kernel_size=1)(noisy_images)
+    x = layers.Concatenate()([x, e])
+
+    skips = []
+    for width in widths[:-1]:
```
[the remark this answers is not in the document]
Comment (beresandras) [reply examples/generative/ddim.py:366]: Done, along with image_size and block_depth.

[the remark this answers is not in the document]
Comment (beresandras) [reply examples/generative/ddim.py:366]: Should embedding_dims and embedding_max_frequency be constructor arguments as well?

File: examples/generative/ddim.py (hunk 6 above)
Comment (fchollet) [reply examples/generative/ddim.py:647]: Got it -- upsampling is not supported with XLA, at least not the op we use. That's too bad.

File: examples/generative/ddim.py [hunk 10]
```diff
@@ -0,0 +1,839 @@
...
+        # this is computationally demanding, kid_diffusion_steps has to be small
+        images = self.denormalize(images)
+        generated_images = self.generate(
+            num_images=batch_size, diffusion_steps=kid_diffusion_steps
+        )
+        self.kid.update_state(images, generated_images)
+
+        return {m.name: m.result() for m in self.metrics}
+
+    def plot_images(self, epoch=None, logs=None, num_rows=3, num_cols=6):
```
[the remark this answers is not in the document]
Comment (fchollet) [reply examples/generative/ddim.py:622]: Ok, sounds good.

File: examples/generative/ddim.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (fchollet) [reply examples/generative/ddim.py:186]: I'd say that's good practice -- maybe just `"tracker"` or something.

File: examples/generative/ddim.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (beresandras) [reply examples/generative/ddim.py:365]: I decided not to put the embedding hyperparams into the constructor, because they are only needed for the Lambda layer for which it is better if they are global.

File: examples/generative/ddim.py (hunk 5 above)
Comment (beresandras) [reply examples/generative/ddim.py]: Added note on using AdamW from tensorflow_addons for tensorflow versions below 2.9

**Comment by beresandras** [comment]
Huggingface demo links added as well.

I believe I am ready, should I update anything else?

**Review by fchollet — APPROVED** [review]
Looks good! Please add the generated files.

**Comment by fchollet** [comment]
Thank you for the great contribution! 👍

**Comment by fchollet** [comment]
Oh, I see the generated files weren't included yet. Please reopen a PR to add them. Sorry about that!

**Comment by beresandras** [comment]
No problem! I added the generated files in #962
